components/script/dom/readablestreamdefaultreader.rs

```diff
@@ -29,135 +28,34 @@ use crate::dom::defaultteereadrequest::DefaultTeeReadRequest;
 use crate::dom::promise::Promise;
 use crate::dom::promisenativehandler::{Callback, PromiseNativeHandler};
-use crate::dom::readablestream::{ReadableStream, get_read_promise_bytes, get_read_promise_done};
+use crate::dom::readablestream::{ReadableStream, get_read_promise_bytes};
 use crate::dom::readablestreamgenericreader::ReadableStreamGenericReader;
 use crate::realms::{InRealm, enter_realm};
 use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
 
 type ReadAllBytesSuccessSteps = dyn Fn(&[u8]);
 type ReadAllBytesFailureSteps = dyn Fn(SafeJSContext, SafeHandleValue);
 
-impl js::gc::Rootable for ReadLoopFulFillmentHandler {}
-
 /// <https://streams.spec.whatwg.org/#read-loop>
-#[derive(Clone, JSTraceable, MallocSizeOf)]
-#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
-struct ReadLoopFulFillmentHandler {
-    #[ignore_malloc_size_of = "Rc is hard"]
-    #[no_trace]
+fn read_loop(
+    reader: &ReadableStreamDefaultReader,
+    cx: SafeJSContext,
     success_steps: Rc<ReadAllBytesSuccessSteps>,
-
-    #[ignore_malloc_size_of = "Rc is hard"]
-    #[no_trace]
     failure_steps: Rc<ReadAllBytesFailureSteps>,
-
-    reader: Dom<ReadableStreamDefaultReader>,
-
-    #[ignore_malloc_size_of = "Rc is hard"]
-    bytes: Rc<DomRefCell<Vec<u8>>>,
-}
-
-impl Callback for ReadLoopFulFillmentHandler {
-    #[cfg_attr(crown, allow(crown::unrooted_must_root))]
-    fn callback(&self, cx: SafeJSContext, v: SafeHandleValue, realm: InRealm, can_gc: CanGc) {
-        let global = self.reader.global();
-        let is_done = match get_read_promise_done(cx, &v, can_gc) {
-            Ok(is_done) => is_done,
-            Err(err) => {
-                self.reader
-                    .release(can_gc)
-                    .expect("Releasing the reader should succeed");
-                rooted!(in(*cx) let mut v = UndefinedValue());
-                err.to_jsval(cx, &global, v.handle_mut(), can_gc);
-                (self.failure_steps)(cx, v.handle());
-                return;
-            },
-        };
-
-        if is_done {
-            // <https://streams.spec.whatwg.org/#ref-for-read-request-close-steps%E2%91%A6>
-            // Call successSteps with bytes.
-            (self.success_steps)(&self.bytes.borrow());
-            self.reader
-                .release(can_gc)
-                .expect("Releasing the reader should succeed");
-        } else {
-            // <https://streams.spec.whatwg.org/#ref-for-read-request-chunk-steps%E2%91%A6>
-            let chunk = match get_read_promise_bytes(cx, &v, can_gc) {
-                Ok(chunk) => chunk,
-                Err(err) => {
-                    //  If chunk is not a Uint8Array object, call failureSteps with a TypeError and abort these steps.
-                    rooted!(in(*cx) let mut v = UndefinedValue());
-                    err.to_jsval(cx, &global, v.handle_mut(), can_gc);
-                    (self.failure_steps)(cx, v.handle());
-                    self.reader
```

Comment:
This release wasn’t needed and caused a crash when I followed the spec correctly.